shared/src/main/scala/execution.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2016-2018 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fetch
+
+import cats.data.NonEmptyList
+import cats.effect._
+import cats.syntax.all._
+
+private object FetchExecution {
+  def parallel[F[_], A](effects: NonEmptyList[F[A]])(
+    implicit CF: ConcurrentEffect[F]
+  ): F[NonEmptyList[A]] =
+    for {
+      fibers <- effects.traverse(CF.start(_))
+      runFibers = fibers.traverse(_.join)
+      _ <- CF.handleErrorWith(runFibers)(error => {
```

Comment:
- We could use `onError { case _ => fibers.traverse(_.cancel) }` instead of `handleErrorWith` + `raiseError`.

- We can also use `traverse_` (or `nonEmptyTraverse_` given that we have a `NonEmptyList`).

- Are we running `runFibers` twice? First in `_ <- CF.handleErrorWIth(...)` and again in `results <- runFibers`?

Maybe
```scala
effects.nonEmptyTraverse(CF.start(_)).flatMap { fibers =>
  fibers
    .nonEmptyTraverse(_.join)
    .onError { case _ => fibers.nonEmptyTraverse_(_.cancel) }
}
```

Reply:
`Fiber` also has an `Applicative` instance, so we could do something like :
```scala
effects.nonEmptyTraverse(Cf.start(_))(CF.compose[Fiber[F, ?]])
```

Not sure if that is better though.